jax_galsim/gsobject.py

```diff
@@ -1032,6 +1157,288 @@ def _drawKImage(
             "%s does not implement drawKImage" % self.__class__.__name__
         )
 
+    @_wraps(_galsim.GSObject._calculate_nphotons)
+    def _calculate_nphotons(self, n_photons, poisson_flux, max_extra_noise, rng):
+        n_photons, g, _rng = calculate_n_photons(
+            self.flux,
+            self._flux_per_photon,
+            self.max_sb,
+            n_photons=n_photons,
+            rng=rng,
+            max_extra_noise=max_extra_noise,
+            poisson_flux=poisson_flux,
+        )
+        if rng is not None:
+            rng._state = _rng._state
+        return n_photons, g
+
+    @_wraps(
+        _galsim.GSObject.makePhot,
+        lax_description="""\
+The JAX-GalSim version of `makePhot`
+
+  - does little to no error checking on the inputs
+  - uses a default of ``n_photons=None`` instead of ``n_photons=0``
+    to indicate that the number of photons should be determined
+    from the flux and gain
+""",
+    )
+    def makePhot(
+        self,
+        n_photons=None,
+        rng=None,
+        max_extra_noise=0.0,
+        poisson_flux=None,
+        photon_ops=(),
+        local_wcs=None,
+        surface_ops=None,
+    ):
+        if surface_ops is not None:
+            from .deprecated import depr
+
+            depr("surface_ops", 2.3, "photon_ops")
+            photon_ops = surface_ops
+
+        if poisson_flux is None:
+            # If n_photons is given, poisson_flux = False
+            poisson_flux = n_photons is None
+
+        if n_photons is not None:
+            # n_photons is the length of an array so it is a python int and
+            # and thus a constant wrt to JIT
+            Ntot = int(n_photons + 0.5)
+            _, g = self._calculate_nphotons(
+                n_photons, poisson_flux, max_extra_noise, rng
+            )
+        else:
+            # here Ntot can be a traced value
+            # one thus must use the fixed_photon_array_size context manager
+            # to ensure that the size of the photon array is fixed if using JIT
+            Ntot, g = self._calculate_nphotons(0.0, poisson_flux, max_extra_noise, rng)
+
+        try:
+            photons = self.shoot(Ntot, rng)
+        except (GalSimError, NotImplementedError) as e:
+            raise GalSimNotImplementedError(
+                "Unable to draw this GSObject with photon shooting.  Perhaps it "
+                "is a Deconvolve or is a compound including one or more "
+                "Deconvolve objects.\nOriginal error: %r" % (e)
+            )
+
+        photons = jax.lax.cond(
```

Comment:
This call saves computation if `g == 1` (which is a common case). The savings may not always be realized because of the ways the underlying compiler can change functions when doing jit/vmap, but that is why the call is used.